### ISSUE TITLE
Severe default PAM configuration issue

### ISSUE THREAD

**Issue by sheckandar** [cockpit-project/cockpit#19008]
### Explain what happens

1. When /etc/cockpit/disallowed-users is missing or has wrong config, root user allowed to login with any wrong password.
2. /etc/cockpit/disallowed-users does not exist on my system for some reason.
3. commenting out `auth       required     pam_listfile.so item=user sense=deny file=/etc/cockpit/disallowed-users onerr=succeed` in /etc/pam.d/cockpit solves the issue.

### Version of Cockpit

286.1 1.el8

### Where is the problem in Cockpit?

None

### Server operating system

Red Hat Enterprise Linux

### Server operating system version

Red Hat Enterprise Linux 8.8 (Ootpa)

### What browsers are you using?

Firefox, Edge

### System log

```shell
2023-06-23 17:03:13 cockpit-session[66578]: pam_unix(cockpit:session): session opened for user root by (uid=0)
2023-06-23 17:03:12 systemd[66609]: pam_unix(systemd-user:session): session opened for user root by (uid=0)
2023-06-23 17:03:11 cockpit-session[66578]: pam_listfile(cockpit:auth): Couldn't open /etc/cockpit/disallowed-users
2023-06-23 17:03:11 cockpit-session[66578]: pam_unix(cockpit:auth): authentication failure; logname= uid=0 euid=0 tty= ruser= rhost=::ffff:192.168.1.100  user=root
2023-06-23 17:03:11 unix_chkpwd[66589]: password check failed for user (root)
```

**Comment by martinpitt** [comment]
> root user allowed to login with any wrong password.

Did that actually work for you? That's **not** what your journal says.

At first sight the pam rule looks fine -- it is `required`, not `sufficient` (which would be a grave bug indeed). 

I tested this on Fedora 38 with cockpit 294, and on RHEL 8.8 with cockpit 286.1. Fedora 38 has "root" in /etc/cockpit/disallowed-users by default, RHEL 8 only has a comment, but no users (for backwards compat reasons).

 * with default /etc/cockpit/disallowed-users, logging in as root with *wrong* password, I get a failure on the login page, and the journal says

> unix_chkpwd[1904]: password check failed for user (root)
> cockpit-session[1902]: pam_unix(cockpit:auth): authentication failure; logname= uid=0 euid=0 tty= ruser= rhost=::ffff:172.27.0.2  user=root

 * with default /etc/cockpit/disallowed-users, logging in as root with *correct* password works:

> cockpit-session[4202]: pam_unix(cockpit:session): session opened for user root by (uid=0)

 * with removing /etc/cockpit/disallowed-users and trying login with *wrong* password, I get
 
> cockpit-session[6514]: pam_unix(cockpit:auth): authentication failure; logname= uid=0 euid=0 tty= ruser= rhost=::ffff:172.27.0.2  user=root
> cockpit-session[6514]: pam_listfile(cockpit:auth): Couldn't open /etc/cockpit/disallowed-users

This looks like your initial log messages, and indeed it says "authententication failure". I don't get the "session opened" messages.

And on Fedora 38, I cannot log in as root even with correct password, as intended.

So again, the question: Can you actually log in as root with a wrong password? If so, I cannot reproduce this, and I'd love to debug that further on your system.

> /etc/cockpit/disallowed-users does not exist on my system for some reason.

Yes, it's only created on new installs, not upgrades.

Thanks!

**Comment by sheckandar** [comment]
> Did that actually work for you? That's not what your journal says.
Yes. I can literally type any password and it lets me in.

I have another box with AlmaLinux 8 installed and there I do not get the same behavior with the same config.
I will try to figure out what specifically causes this behavior. I could've been hacked, for all I know.

**Comment by sheckandar** [comment]
OK Got some more info.

Looks like my password-auth file had some issues:
```

auth        required                                     pam_env.so
auth        required                                     pam_faildelay.so delay=2000000
auth        sufficient                                   pam_winbind.so krb5_auth try_first_pass cached_login
auth        sufficient                                   pam_unix.so nullok
auth        [default=1 ignore=ignore success=ok]         pam_usertype.so isregular
auth        required                                     pam_deny.so
```
I don't remember why I set it that way, but the third line was the cause of of it all.

After comparing the file to the one on AlmaLinux box, I moved the offending line to line 5 and the issue went away.

auth        required                                     pam_env.so
auth        required                                     pam_faildelay.so delay=2000000
auth        sufficient                                   pam_unix.so nullok
auth        [default=1 ignore=ignore success=ok]         pam_usertype.so isregular
auth        sufficient                                   pam_winbind.so krb5_auth try_first_pass cached_login
auth        required                                     pam_deny.so
